### PR TITLE
partition: fix exchange partition validition check for null value

### LINKED ISSUE #47167: Exchange List Partition add validition check  tests for null value handling and default partition
Please create a followup issue, to test null handling in validation, since I assume there may be hidded issues in non-list cases?

_Originally posted by @mjonss in https://github.com/pingcap/tidb/pull/46021#pullrequestreview-1636530327_

### PR DESCRIPTION
**Pull request by jiyfhust**
### What problem does this PR solve?

Issue Number: close #47167

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by jiyfhust** [comment]
The "partition by range columns(a, b, c)" case is not fixed by this pr, it seems difficult to fix it directed in func buildCheckSQLConditionForRangeColumnsPartition.

Mayby we can handle it this way:
```
CREATE TABLE t1 (a int, b int, c int) partition by range columns(a, b, c)(
		partition p0 values less than (100, 1000, 10000), 
		partition p1 values less than (200, 2000, 20000),
		partition p2 values less than (maxvalue, null, null));
CREATE TABLE t2 (a int, b int, c int);

alter table t1 EXCHANGE PARTITION p0 WITH TABLE t2; 

the checksql:

select 1 from t2 where (a, b, c) >= (100, 1000, 10000) imit 1;

alter table t1 EXCHANGE PARTITION p1 WITH TABLE t2; 

the checksql:

select 1 from t2 where (a, b, c) < (100, 1000, 10000) or (a, b, c) >= (200, 2000, 20000) or a is null or a = 100 and b is null or c = 100 and b = 1000 and c is null limit 1;

alter table t1 EXCHANGE PARTITION p2 WITH TABLE t2; 

the checksql:
select 1 from t2 where (a, b, c) < (200, 2000, 20000) or a is null or a = 200 and b is null or c = 200 and b = 2000 and c is null limit 1;

```

**Comment by mjonss** [comment]
> The "partition by range columns(a, b, c)" case is not fixed by this pr, it seems to be difficult to fix it directly in func [buildCheckSQLConditionForRangeColumnsPartition](https://github.com/pingcap/tidb/blob/34438f84a1fb0ac1220e01028175361e0137eb22/ddl/partition.go#L3587-L3600).
> ...

I thing you are on the right path, [here](https://github.com/pingcap/tidb/blob/34438f84a1fb0ac1220e01028175361e0137eb22/planner/core/expression_rewriter.go#L275) is an example how an expression can be rewritten.

Please feel free to split out the RANGE COLUMS case from this PR.

File: ddl/partition.go [hunk 1]
```diff
@@ -3572,13 +3580,17 @@ func buildCheckSQLConditionForRangeExprPartition(pi *model.PartitionInfo, index
 		paramList = append(paramList, driver.UnwrapFromSingleQuotes(pi.Definitions[index].LessThan[0]))
 	} else if index == len(pi.Definitions)-1 && strings.EqualFold(pi.Definitions[index].LessThan[0], partitionMaxValue) {
 		buf.WriteString(pi.Expr)
-		buf.WriteString(" < %?")
+		buf.WriteString(" < %? or ")
+		buf.WriteString(pi.Expr)
+		buf.WriteString(" is null")
```
Comment (mjonss) [inline ddl/partition.go:3585]: Can there be a case where the expression string also would need to be wrapped into parentheses?

Comment (jiyfhust) [reply ddl/partition.go:3585]: I have to say i'm not sure about it. I tested `partition by range (id+1)`, the expression is already wapped. And if `pi.Expr is null` needed, then `pi.Expr < %?` seems also needs. I think we can keep it now before we can make sure some cases as you described exist.

**Comment by jiyfhust** [comment]
@mjonss I think maybe we can review this pr as present.

**Review by mjonss — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Comment by jiyfhust** [comment]
PTAL @ti-chi-bot

**Comment by jiyfhust** [comment]
@jebter @aytrack Please remove the `needs-rebase` lable, it can't be auto removed by ti-chi-bot.

**Comment by jebter** [comment]
> @jebter @aytrack Please remove the `needs-rebase` lable, it can't be auto removed by ti-chi-bot.

@wuhuizuo PTAL

**Comment by jiyfhust** [comment]
> please assign [wjhuang2016](https://github.com/wjhuang2016) for approval

/cc @wjhuang2016

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**